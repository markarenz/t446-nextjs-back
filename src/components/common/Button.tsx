import styles from '../../styles/modules/Button.module.scss';

type Props = {
  children: JSX.Element;
  variant?: string;
  color?: string;
  size?: string;
  onClick?: Function;
  disabled: boolean;
  tabIndex?: number;
};

const Button: React.FC<Props> = ({
  children,
  variant,
  color,
  size,
  onClick,
  disabled,
  tabIndex
}) => {
  const classSize = size === 'small' ? styles.small : '';
  const classVariant =
    variant === 'outlined' ? styles.outlined : styles.contained;
  let classColor = styles.colorPrimary;
  if (color === 'secondary') {
    classColor = styles.colorSecondary;
  }
  return (
    <button
      onClick={!!onClick ? (e) => onClick(e) : () => {}}
      type="button"
      tabIndex={tabIndex}
      disabled={disabled}
      className={`${styles.button} ${classSize} ${classVariant} ${classColor}`}
    >
      {children}
    </button>
  );
};

export default Button;
