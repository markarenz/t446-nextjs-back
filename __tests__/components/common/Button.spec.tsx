import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Button from '../../../src/components/common/Button';

const defaultProps = {
  title: 'Test',
  variant: 'contained',
  color: 'primary',
  onClick: jest.fn(),
  size: 'large',
  disabled: false
};
const altProps = {
  ...defaultProps,
  variant: 'outlined',
  size: 'small',
  color: 'secondary',
  onClick: undefined
};
describe('Button', () => {
  it('renders default button', async () => {
    render(
      <Button {...defaultProps}>
        <span>Test</span>
      </Button>
    );
    const element = await screen.findByRole('button');
    expect(element).toMatchSnapshot();
  });
  it('renders button with secondary color and outlines', async () => {
    render(
      <Button {...altProps}>
        <span>Test</span>
      </Button>
    );
    const element = await screen.findByRole('button');
    expect(element).toBeInTheDocument();
  });
  it('fires onClick when clicked', async () => {
    render(
      <Button {...defaultProps}>
        <span>Test</span>
      </Button>
    );
    const element = await screen.findByRole('button');
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );
    expect(defaultProps.onClick).toHaveBeenCalled();
  });

  it('does not fire onClick when no onClick prop provided', async () => {
    render(
      <Button {...altProps}>
        <span>Test</span>
      </Button>
    );
    const element = await screen.findByRole('button');
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );
    expect(defaultProps.onClick).toHaveBeenCalled();
  });
});
