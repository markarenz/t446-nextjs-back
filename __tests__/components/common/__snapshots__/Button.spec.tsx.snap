// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders default button 1`] = `
<button
  class="button  contained colorPrimary"
  role="button"
  type="button"
>
  <span>
    Test
  </span>
</button>
`;
